### ISSUE TITLE
Parser error for generic arrow functions in very specific circumstances

### ISSUE THREAD

**Issue by kgram** [microsoft/TypeScript#48154]
# Bug Report

### 🔎 Search Terms

parser, generic

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about **generic arrow functions**

JSX is disabled, so this is not related to JSX parsing.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0#code/PTAEEEAdIQwJwKYDsAupZwM4LqHcB7XAdwAtlQAzASyzUkMhxQE9QDLREUBXOJBABN2AIwBWCAMZpqmUAHNkOapKo8k06gSShi1FKVCCElGDwA2aVkwCwAKEnbMaU9XOYAjKAC8oABQAlD4AfP4A3vagoCgIzgBcoAA8ACo+oOrGNAKCwX4AbjDmPAgJyQGlIaAFRQgANJF4AB4ocDAAaoXFCYGVYQC+9XZ9Afb2IKAA6vqkBDxoMEYmZpbRLEy1oPq6RADWclkI9o5Izttwe16+Pd6hfhF2UTHxScm51V2gZRU3VZ11DQhmq0OjVukEfv1BsNRnZxuA5NhjsIGAQmHBWBstsRdvtqAIjk40NjzpgAExpa63e5RQEtdp-MG9AYNJ4oBIpNIZEx4oRvBmfcqfSrvf5DEZ2MZgZLkKi0U4otGsUAAfQAtjxnMrQCIEKAFpR1JptBsDLN5KQCScXDA3GSKeCqQCgfTQaAAORuwaPWJsl6cpCZHk5fL8r5Cn4iqEBIA)

### 💻 Code

```ts
// Apparent parser error when first property of returned object is generic function with default type
const fails1 = () => ({
  test: <T = undefined>(value: T): T => value,
  extraValue: () => {},
})

// Without a default type, it works fine
const works1 = () => ({
  test: <T>(value: T): T => value,
  extraValue: () => {},
})

// As second property, it works fine
const works2 = () => ({
  extraValue: () => {},
  test: <T = undefined>(value: T): T => value,
})

// The first property _must_ be a function, though
const fails2 = () => ({
  extraValue: '',
  test: <T = undefined>(value: T): T => value,
})
```

### 🙁 Actual behavior

In very specific circumstances, the parser seems to produce nonsense. The two failing examples produce respectively 15 and 23 errors.

As far as I can tell, it only happens when a function returns an object where the first arrow-function property is generic with a default type parameter. If there is no default type, it works fine. If another arrow-function property proceeds the generic one, it works fine (simply reordering the lines in the example can remove the problem). Any non-arrow-function properties before the function does not affect the error, but has nonsense errors themselves.

### 🙂 Expected behavior

No errors in any of the examples.

**Comment by AllySummers** [comment]
Just a random observation - if you add a comma after `T = undefined`, to be `T = undefined,` then it works without errors ([playground link](https://www.typescriptlang.org/play?jsx=0&ssl=10&ssc=24&pln=11&pc=3&filetype=ts#code/PTAEEEAdIQwJwKYDsAupZwM4LqHcB7XAdwAtlQAzASyzUkMhxQE9QDLREUBXOJBABN2AIwBWCAMZpqmUAHNkOapKo8k06gSShi1FKVCCElGDwA2aVkwCwAKEnbMaU9XOYAjKAC8oABQAlD4AfP4A3vagoCgIzgBcoAA8ACo+oOrGNAKCADTBfgBuMOY8CAnJAeUhoEUlCDmReAAeKHAwAGrFpQmB1WEAvg12-QH29iCgAOr6pAQ8aDBGJmaW0SxMOaD6ukQA1nJZCPaOSM47cPtevr3eoX4RdlEx8UnJ+bXdoBVVtzVd9Y0EC02p06j0gr8BkMRmM7BNwHJsCdhAwCEw4KxNttiHsDtQBMcnGgcRdMAAmNI3O4PKJA1odf7gvqDRrPFAJFJpDImfFCd6Mr6VL7VD4A4ajOzjMDJchUWhnVHo1igAD6AFseM4VaARAhQItKOpNNpNgY5vJSITTi4YG5yZSIdTAcCGWDQAByd1DJ6xdmvLlITK83L8t3fYW-UXQgJAA)). Not suggesting as a workaround, just something I noticed (and thought to try after having weird times with generics in TSX files)

**Comment by Oblosys** [comment]
The error also happens on (half) parenthesized objects. Both `fails1` & `fails2` show the same parse errors, whereas `ok` is fine.

```ts
const fails1 = ({test: <T = 0>(value: T) => 0}) // ERRORS: Duplicate identifier '(Missing)'. etc.
const fails2 = ({test: <T = 0>(value: T) => 0}  // ERRORS: Duplicate identifier '(Missing)'. etc.
const ok =      {test: <T = 0>(value: T) => 0}  // Fine, despite missing `)` above
```

[Playground Link](https://www.typescriptlang.org/play?jsx=0&ts=4.6.2&filetype=ts#code/MYewdgzgLgBAZgQwJYBsIEYYF4YAoDeUAptAFwwA8AKtjAAwB8uAbgigK5HlUCU2D9AL58A9CJgBRAEpSA8lIDK5ACLsADiiTAExGEgAmRMFCRwkRAE4wA5LgCySCBCRgA5j2sA6GESjBPALAAUKCQsIioEABMtATEZJQ0OIwsbJzcfFgCdIIwMGKSMvJKMKoaWjpEeobGpuZWtg5OLu5ePn6BIeDQMCAA1rR5Q4QkUOTUtCmsHFwwvPxCeQUAYi5EADQwhhBqSLoAto7ObjAABjynMAgARiDMRMFAA)

**Comment by jakebailey** [comment]
This is fixed by #48493, which means that this is similar to #32914.
